File: Python/Algorithm/binary_search.py
def binary_search(li: list[int], item: int) -> int:
    l = 0
    r = len(li) - 1

    while l <= r:
        mid = l + (r - l) // 2

        if li[mid] == item:
            return mid
        elif li[mid] > item:
            r = mid - 1
        else:
            l = mid + 1

    return -1


li = [1, 2, 3, 4, 5]
item = 3
result = binary_search(li, item)

print(result)

File: Nim/Recursions/BinarySearch.nim
proc binary_search(list: seq[int], item: int, l, r: int): int =
    if l <= r:
        var mid: int = l + (r - l) div 2

        if list[mid] == item:
            return mid
        elif list[mid] > item:
            return binary_search(list, item, l, mid-1)
        else:
            return binary_search(list, item, mid+1, r)
    return -1


when isMainModule:
    var list: seq[int] = @[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    let item: int = 4
    var l: int = 0
    var r: int = list.len - 1


    echo binary_search(list, item, l, r)